docs/specs/saml.md

```diff
@@ -0,0 +1,701 @@
+
+The hook is expected to return a object, which the key is the `Name` of the SAML attribute, and the value is the value of the SAML attribute.
+
+Here is an typescript hook example:
+
+```typescript
+import {
+  SAMLAttributesMappingsPayload,
+  SAMLAttributesMappingsResponse,
+} from "https://deno.land/x/authgear-deno-hook@0.1.0/mod.ts";
+
+export default async function (
+  e: SAMLAttributesMappingsPayload
+): Promise<SAMLAttributesMappingsResponse> {
+  return {
+    "https://schemas.xmlsoap.org/ws/2005/05/identity/claims/email":
+      e.user.email,
+    "https://schemas.xmlsoap.org/ws/2005/05/identity/claims/upn":
+      e.user.preferred_username,
+  };
+}
+```
+
+## <a id="5"></a> Bindings
+
+### <a id="5_1"></a> Supported Bindings
+
+The following bindings are supported:
+
+- urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect
+- urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Post
+
+### <a id="5_2"></a> Unsupported Bindings
+
+The following bindings are not supported at the moment, but we would like to support it in the future:
+
+- urn:oasis:names:tc:SAML:2.0:bindings:SOAP
+
+  - SAML SOAP Binding is a synchronous binding which can be used in the single logout profile to send `<LogoutRequest>` to other session participants.
+  - Many other profiles can be supported if we support SOAP binding. See [SAMLConformance](https://groups.oasis-open.org/higherlogic/ws/public/download/35393/sstc-saml-conformance-errata-2.0-wd-04-diff.pdf) Section 2.
+
+We have no plan to support other bindings not mentioned at the moment.
+
+## <a id="6"></a> Signing
+
+- Authgear as an IdP always sign all issued requests, responses and assertions.
+  - The signatures are created according to [SAMLCore](https://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf) Section 5.
+  - Each project will have independent signing certs. New certs can be generated by the portal. To facilitate cert rotations, a maximum of two certs is allowed. However, only one single cert can be used in signing. The active signing cert is controlled by the config `saml.signing.key_id`.
+  - Signing Algorithm ([SignatureMethod](https://www.w3.org/TR/xmldsig-core1/#sec-SignatureMethod)) is configurable with config `saml.signing.signature_method`. Please see below for supported values.
+  - Digest Method ([DigestMethod](https://www.w3.org/TR/xmldsig-core1/#sec-DigestMethod)) is configurable with config `saml.signing.digest_method`. Please see below for supported values.
+- If Service Providers has signing certs uploaded, all requests and responses from the Service Provider will be validated against the provided certs.
+  - To facilitate cert rotations, two certs are allowed to be uploaded. When there are two certs, the requests and responses will be considered valid if its signature can be validated by any one of the uploaded certs.
+
+### <a id="6_1"></a> Configs
+
+```yaml
+saml:
+  signing:
+    key_id: key01
+    signature_method: RSAwithSHA256
+    digest_method: SHA256
```

Comment:
The library we are using ("github.com/russellhaering/goxmldsig") does not support configuring digest method and signature method separately. Maybe we just remove `digest_method` and use `signature_method` to control both.

- [ ] Remove `digest_method`

Reply:
Maybe replace `signature_method` and `digest_method` with `signature_algorithm`. Signature algorithm should be a more common word for this thing. Reference https://datatracker.ietf.org/doc/html/rfc7517#section-4.4

For the values, does the SAMLCore spec define any signature algorithm constants? If not, maybe we can use the values from JWA. https://www.rfc-editor.org/rfc/rfc7518.html#section-3.1 I think we support `RS256` only.

Reply:
We can rename the config key to signature_algorithm.
The value `RSAwithSHA256` is from https://www.w3.org/TR/xmldsig-core/ section 6.

Reply:
It seems that the identifier is `http://www.w3.org/2001/04/xmldsig-more#rsa-sha256`, rather than `RSAwithSHA256`

Reply:
https://datatracker.ietf.org/doc/html/rfc6931

Reply:
`RSAwithSHA256` is the name mentioned in section 6.1. The URI is indeed `http://www.w3.org/2001/04/xmldsig-more#rsa-sha256`. We can change to use the URI instead.